File: linear_regression/linear.py
type Vector = list[float]
type Matrix = list[Vector]


def _dot(a: Vector, b: Vector) -> float:
        if len(a) != len(b):
            raise ValueError("Error: vectors must have same size")

        res = 0.0

        for i in range(len(a)):
            res += a[i] * b[i]

        return res

class LinearRegression:
    def __init__(self, learning_rate: float = 0.01) -> None:
        self.weights: Vector = []
        self.bias = 0.0
        self.learning_rate = learning_rate

    def _validate_data(self, x: Matrix, y: Vector) -> None:
        if len(x) == 0:
            raise ValueError("x must not be empty")

        if len(x) != len(y):
            raise ValueError("x and y must contain the same number of samples")

        feature_count = len(x[0])

        if feature_count == 0:
            raise ValueError("samples must contain at least one feature")

        for sample in x:
            if len(sample) != feature_count:
                raise ValueError("all samples must have the same number of features")

    def _predict_one(self, x: Vector) -> float:
        return _dot(self.weights, x) + self.bias

    def predict(self, x: Matrix) -> Vector:
        prediction: Vector = []

        for sam in x:
            prediction.append(self._predict_one(sam))

        return prediction

    def _mse(self, x: Matrix, y: Vector) -> float:
        self._validate_data(x, y)

        score = 0.0

        for i in range(len(x)):
            prediction = self._predict_one(x[i])
            error = prediction - y[i]
            score += error**2

        return score / len(x)

    def mse(self, x: Matrix, y: Vector) -> float:
        return self._mse(x, y)

    def _gradient_weights(self, x: Matrix, y: Vector) -> Vector:
        self._validate_data(x, y)

        grad: Vector = [0.0] * len(self.weights)

        for i in range(len(x)):
            prediction = self._predict_one(x[i])
            error = prediction - y[i]

            for j in range(len(self.weights)):
                grad[j] += error * x[i][j]

        for j in range(len(grad)):
            grad[j] *= 2.0 / len(x)

        return grad

    def gradient_w(self, x: Matrix, y: Vector) -> Vector:
        return self._gradient_weights(x, y)

    def _gradient_bias(self, x: Matrix, y: Vector) -> float:
        self._validate_data(x, y)

        gradient_sum = 0.0

        for i in range(len(x)):
            pred = self._predict_one(x[i])
            error = pred - y[i]
            gradient_sum += error

        return 2.0 * gradient_sum / len(x)

    def gradient_b(self, x: Matrix, y: Vector) -> float:
        return self._gradient_bias(x, y)

    def train(self, x: Matrix, y: Vector, epochs: int = 1_000) -> None:
        self._validate_data(x, y)

        feature_count = len(x[0])
        self.weights = [0.0] * feature_count
        self.bias = 0.0

        for _ in range(epochs):
            gradient_w = self._gradient_weights(x, y)
            gradient_b = self._gradient_bias(x, y)

            for j in range(len(self.weights)):
                self.weights[j] -= self.learning_rate * gradient_w[j]

            self.bias -= self.learning_rate * gradient_b

File: linear_regression/c/linear.c
#include "linear.h"
#include <stdlib.h>
#include <math.h>

static double dot(const double *a, const double *b, size_t size) {
    double res = 0;

    for(size_t i = 0; i < size; ++i) {
        res += a[i] * b[i];
    }

    return res;
}

int linear_init(
    LinearRegression *model,
    size_t feature_cnt,
    double learning_rate
) {
    if(model == NULL || feature_cnt == 0 || learning_rate < 0.0) {
        return 0;
    }

    model -> weigths = calloc(feature_cnt, sizeof(double));

    if (model -> weigths == NULL) {
        return 0;
    }

    model -> bias = 0.0;
    model -> learning_rate = learning_rate;
    model -> feature_cnt = feature_cnt;

    return 1;
}

void linear_free(LinearRegression *model) {
    if (model == NULL) {
        return;
    }

    free(model->weigths);
    model -> weigths = NULL;
    model -> feature_cnt = 0;
}

double learning_predict_one(const LinearRegression *model, const double *sample) {
    return dot(model -> weigths, sample, model -> feature_cnt) + model->bias;
}

double learning_mse(
    const LinearRegression *model,
    const double *x,
    const double *y,
    size_t sample_cnt
) {
    if (model == NULL || model -> weigths || x == NULL || y == NULL || sample_cnt == 0) {
        return NAN;
    }

    double score = 0.0;

    for(size_t i = 0; i < sample_cnt; ++i) {
        const double *sample = &x[i * model -> feature_cnt];

        double predict = learning_predict_one(model,sample);
        double err = predict - y[i];

        score += err * err;
    }

    return score / (double)sample_cnt;
}

int learn_train(
    LinearRegression *model,
    const double *x,
    const double *y,
    size_t sample_cnt,
    size_t epochs
) {
    if (model == NULL ||
        epochs == 0 ||
        sample_cnt == 0 ||
        model -> weigths == NULL ||
        x == NULL ||
        y == NULL ||
        model -> feature_cnt == 0
    ) return 0;


    double *gradient_w = malloc(model->feature_cnt * sizeof(double));

    if(gradient_w == NULL) return 0;

    for(size_t epo = 0; epo < epochs; ++epo) {
        for(size_t j = 0; j < model->feature_cnt; ++j){
            gradient_w[j] = 0.0;
        }

        double gradient_bias = 0.0;

        for(size_t i = 0;i<sample_cnt;++i) {
            const double *sample = &x[i * model->feature_cnt];

            double pred = learning_predict_one(model, sample);

            double err = pred - y[i];

            for(size_t j = 0; j < model -> feature_cnt;++j) {
                gradient_w[j] = err * sample[j];
            }
            gradient_bias += err;
        }
        double scale = 2.0 / (double)sample_cnt;

        for(size_t j =0; j<model->feature_cnt;++j) {
            model->weigths[j] -= model->learning_rate * scale * gradient_w[j];
        }

        model->bias -= model->learning_rate * scale * gradient_bias;
    }

    free(gradient_w);

    return 1;
}
